### ISSUE TITLE
Budget seems to improve performance up to a point and then nose dives

### ISSUE THREAD

**Issue by bjoernholzhauer** [perpetual-ml/perpetual#19]
I've noticed in at least two Kaggle Playground Series competitions that as budget improves performance improves, but at some point (usually `budet=2.1` is still good) the performance deteroriates noticably.

See e.g. in the current playground, where in [this notebook](https://www.kaggle.com/code/bjoernholzhauer/ps-s04e10-perpetualbooster) budget 2.1 (see notebook version 14) is pretty good (mean ROC AUC of 0.9535 also 0.95543 on the public leaderboard), but with `budget=3` you get a mean cross-validated ROC AUC of 0.86384 (0.86957 on the public leaderboard).

See also [this discussion](https://www.kaggle.com/competitions/playground-series-s4e10/discussion/541573 ).

I also had a similar thing on the previous competition (e.g. look at [this notebook](https://www.kaggle.com/code/bjoernholzhauer/ps-s04e09-used-car-prices-perpetualbooster/) and see what happens when you vary `budget`).

I'm a little unsure what is going on. Is this overfitting? Numerical issues? Something else?

**Comment by deadsoul44** [comment]
Hello,

The number of boosting rounds is internally limited to 1000. The algorithm might be hitting this limit with very high budget (3.0). The algorithm should print a warning if it hits this number but I couldn't see the warning (it might be due to logging settings of kaggle). Could you please check `model.number_of_trees`?

For practical real world cases, there shouldn't be need for a budget more than `2.0`. That is why the number of boosting rounds is limited to 2.0.

Another note: There should be no need for cross validation.

**Comment by bjoernholzhauer** [comment]
Looking at the current tabular playground competition, it looks like `budget=2.1` already hits 1000 trees. `budget=3.0` also has 1000 trees. See[this notebook](https://www.kaggle.com/code/bjoernholzhauer/ps-s04e10-perpetualbooster-single-fit).

I also cross-validated performance vs. budget:

| Budget | Mean Number of Trees Across Folds | Mean ROC AUC |
|--------|-----------------------------------|--------------|
| 0.1    | 53                              | 0.9434       |
| 0.2    | 50                             | 0.9463       |
| 0.3    | 47                              | 0.9506       |
| 0.4    | 44                              | 0.9507       |
| 0.5    | 49                              | 0.9519       |
| 0.75  | 70                             | 0.9539        |
| 1.0    | 98                            | 0.9541       |
| 1.25  | 156                               | 0.9537   |
| 1.5    | 248                               | 0.9537   |
| 1.75  | 384                               | 0.9536   |

Certainly, performance stops to go up at some point and we just get more trees that seem to make little difference. Except for this purely empirical observation on one dataset (from my side, at least), I'm unsure why a limit of 1000 trees makes sense. In LightGBM/CatBoost etc. we frequently use more iterations than that. I don't have that good a feel for PerpetualBooster, but I wonder whether it is obviously true that 1000 is enough (especially when there's a clear performance gap to the other gradient boosted decision trees).

Why would one not need cross-validation? How do I know how the model is likely to perform? The cross-validation nicely shows the drop in performance with `budget=3` in the example. How do I know how to stack it (on out-of-fold predictions) with other models?

**Comment by deadsoul44** [comment]
For testing purposes, cross validation is useful. I meant the algorithm has a built-in mechanism to check generalization as described in the blog post. 

PerpetualBooster provides better performance than LightGBM. We also have promising results in AutoML benchmark. We will publish the results in the paper. 

As seen in the results, the performance peaked at `budget=1.0` with `98` boosting rounds. So, in most of the practical cases, there shouldn't be need for more 1000 boosting rounds. But, we might provide this as a parameter like `iteration_limit`. We also have a `timeout` parameter as an additional info.

**Comment by deadsoul44** [comment]
Hello, `v0.6.0` is released with `iteration_limit` option. It is still 1000 by default but it can be increased to experiment with very large budget (>2.0).
